stdlib/fun.ml

```diff
@@ -17,3 +17,27 @@ external id : 'a -> 'a = "%identity"
 let const c _ = c
 let flip f x y = f y x
 let negate p v = not (p v)
+
+exception Finally_raised of exn
+
+let protect ~(finally : unit -> unit) work =
+  let finally_no_exn () =
+    try finally () with e -> raise (Finally_raised e)
+  in
+  match work () with
+  | result -> finally_no_exn () ; result
+  | exception work_exn ->
+      let work_bt = Printexc.get_raw_backtrace () in
+      finally_no_exn () ;
+      Printexc.raise_with_backtrace work_exn work_bt
+
+let with_resource ~acquire ~(release : _ -> unit) work =
+  let release_no_exn release resource =
+    try Printexc.catch release resource
+    with _ -> exit 2 (* printing error message failed *)
```

Comment:
I'm unconvinced by the `exit` design (also please always leave such low-level exit codes for application exit semantics). Why not simply raise `Release_raised` with the exception ? 

This would still convey the idea that some "invariants about the state of the world" have been violated but will avoid giving us terrible toplevel usability where `Ctr-c`ing a computation might kill your toplevel session because an underlying library might be using `Fun.with_resource` when you `Ctrl-c`.

Reply:
> please always leave such low-level exit codes for application exit semantics
I have seen your suggestion to use 125 elsewhere, but on the other hand the use of 2 by the runtime and the libraries is consistent and documented.
> Why not simply raise Release_raised with the exception ?
See below
> avoid giving us terrible toplevel usability where `Ctr-c`ing a computation might kill your toplevel session
With a correct `with_resource` implementation, this does not happen because of masking. For the previous PR where the implementation in bytecode was known to be incorrect, don't worry, I had temporary workarounds for Ctrl+C in the toplevel in mind, but it would have been premature to propose them.

Reply:
Closing because outdated.